src/Microsoft.OData.Core/JsonLight/ODataJsonLightWriter.cs

```diff
@@ -1003,6 +1003,17 @@ protected override void EndNestedResourceInfoWithContent(ODataNestedResourceInfo
                     this.jsonWriter.EndArrayScope();
                 }
             }
+            else
+            {
+                JsonLightNestedResourceInfoScope navigationLinkScope = (JsonLightNestedResourceInfoScope)this.CurrentScope;
+
+                // If we wrote entity reference links for a collection navigation property but no
+                // resource set afterwards, we have to now close the array of links.
+                if (navigationLinkScope.EntityReferenceLinkWritten && !navigationLinkScope.ResourceSetWritten && nestedResourceInfo.IsCollection.Value)
+                {
+                    this.jsonWriter.EndArrayScope();
+                }
+            }

```

Comment:
Why don't we just move this before line 987, and remove 989-997 (and this else condition)?

Reply:
Good. That's more clear. #Resolved.

---
In reply to: [279949030](https://github.com/OData/odata.net/pull/1459#discussion_r279949030) [](ancestors = 279949030)